### PR TITLE
Hw17

### PR DESCRIPTION
**Pull request by voloshynamaryna11**

### CODE REVIEW

File: src/main/java/core/basesyntax/BadOperandException.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax;
+
+public class BadOperandException extends Exception {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/BadOperandException.java:3]: let's extend RuntimeException - we wouldn't need to catch it

File: src/main/java/core/basesyntax/HelloWorld.java [hunk 2]
```diff
@@ -1,8 +1,36 @@
 package core.basesyntax;
 
-/**
- * Feel free to remove this class and create your own.
- */
 public class HelloWorld {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/HelloWorld.java:3]: please give this class a relevant name

File: src/main/java/core/basesyntax/HelloWorld.java [hunk 3]
```diff
@@ -1,8 +1,36 @@
...
+    public double calculate(double value1, double value2, char operation)
+            throws BadOperandException {
 
+        double result;
+
+        switch (operation) {
+            case ('*'):
+                result = value1 * value2;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/HelloWorld.java:11]:
```suggestion
                return value1 * value2;
```
you don't really need `result` variable

File: src/main/java/core/basesyntax/HelloWorld.java [hunk 4]
```diff
@@ -1,8 +1,36 @@
...
+                break;
+            case ('/'):
+                if (value2 == 0) {
+                    throw new ArithmeticException("You can't divide by zero");
+                }
+                result = value1 / value2;
+                break;
+            case ('^'):
+                if (value1 == 0 && value2 < 0) {
+                    throw new ArithmeticException("You can't divide by zero");
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/HelloWorld.java:27]: I think this message does not fit here

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 5]
```diff
@@ -1,8 +1,80 @@
 package core.basesyntax;
 
-/**
- * Feel free to remove this class and create your own.
- */
+import org.junit.*;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:3]: replace this with separate imports

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 6]
```diff
@@ -1,8 +1,80 @@
...
+
 public class HelloWorldTest {
+    private HelloWorld helloWorld;
+    private final static double SUM_OF_TWO_POSITIVE_NUMBERS = 46.0;
+    private final static double SUM_OF_TWO_POSITIVE_AND_NEGATIVE_NUMBERS = 0;
+    private final static double ResultOfDividingByNotNull = 45 / 7.0;
+    private final static double ResultBySubtractionOfTwoNegativeNumbers = 5;
+    private final static double ResultWithMultiplicationWithNull = 0;
+    private final static double ResultWithNullBasisAndNullDegree = 1;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:12]: are these constants?

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 7]
```diff
@@ -1,8 +1,80 @@
...
+
+    @Before
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:14]: let's use @BeforeClass
it will allow us to create only one Calculator instance

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 8]
```diff
@@ -1,8 +1,80 @@
...
+    public void setUp() {
+        helloWorld = new HelloWorld();
+    }
+
+    @After
+    public void tearDown() {
+        helloWorld = null;
+    }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:22]: you won't need this

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 9]
```diff
@@ -1,8 +1,80 @@
...
+
+    @Test
+    public void getSumOfTwoPositiveNumbers() throws BadOperandException {
+        double actualResult = helloWorld.calculate(12.5, 33.5, '+');
+        double expectedResult = SUM_OF_TWO_POSITIVE_NUMBERS;
+        Assert.assertEquals(expectedResult, actualResult, 0);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:28]: do you need `0` here?
let's make a constant field `DELTA` and use it where necessary

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 10]
```diff
@@ -1,8 +1,80 @@
...
+    }
+
+    @Test
+    public void getSumOfTwoNegativeAndPositiveNumbers() throws BadOperandException {
+        double actualResult = helloWorld.calculate(58, -58, '+');
+        double expectedResult = SUM_OF_TWO_POSITIVE_AND_NEGATIVE_NUMBERS;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:34]: let's remove unnecessary variables
you can use your constant in `Assert.assertEquals()`

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
also, I think you should add more test cases for for positive, negative and zero input for different operations
you can use plain result values, there is no need to put them as constants

File: src/main/java/core/basesyntax/HelloWorld.java (hunk 4 above)
Comment (voloshynamaryna11) [reply src/main/java/core/basesyntax/HelloWorld.java:27]: так вознесение в отрицательную степень это и есть деление

File: src/main/java/core/basesyntax/BadOperandException.java (hunk 1 above)
Comment (voloshynamaryna11) [reply src/main/java/core/basesyntax/BadOperandException.java:3]: done

File: src/main/java/core/basesyntax/HelloWorld.java (hunk 2 above)
Comment (voloshynamaryna11) [reply src/main/java/core/basesyntax/HelloWorld.java:3]: done

File: src/main/java/core/basesyntax/HelloWorld.java (hunk 3 above)
Comment (voloshynamaryna11) [reply src/main/java/core/basesyntax/HelloWorld.java:11]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 6 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:12]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 5 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:3]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 7 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:14]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 8 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:22]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 9 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:28]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 10 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java:34]: done

File: src/main/java/core/basesyntax/Calculator.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package core.basesyntax;
+
+public class Calculator {
+    public double calculate(double value1, double value2, char operation)
+            throws BadOperandException {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Calculator.java:5]:
```suggestion
    public double calculate(double value1, double value2, char operation) {
```

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 12]
```diff
@@ -1,8 +1,73 @@
...
-/**
- * Feel free to remove this class and create your own.
- */
+import org.junit.Test;
+import org.junit.BeforeClass;
+import org.junit.Assert;
+
+
 public class HelloWorldTest {
+    private static Calculator calculator;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:9]: it is good practice to place constants first, than all other fields

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
add at least 5 more test cases with positive, negative and zero number combinations for different operations

File: src/main/java/core/basesyntax/Calculator.java (hunk 11 above)
Comment (voloshynamaryna11) [reply src/main/java/core/basesyntax/Calculator.java]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 12 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 13]
```diff
@@ -1,8 +1,115 @@
...
 
-/**
- * Feel free to remove this class and create your own.
- */
+import org.junit.Test;
+import org.junit.BeforeClass;
+import org.junit.Assert;
+
+
 public class HelloWorldTest {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:8]: change class name to more appropriate

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 14]
```diff
@@ -1,8 +1,115 @@
...
+    private final static double DELTA = 0.001;
+    private static Calculator calculator;
+
+    @BeforeClass
+    public static void setUp() {
+        calculator = new Calculator();
+    }
+
+    @Test
+    public void getSumOfTwoPositiveNumbers() throws BadOperandException {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:19]:
```suggestion
    public void getSumOfTwoPositiveNumbers() {
```
fix this in all methods

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 15]
```diff
@@ -1,8 +1,115 @@
...
+        Assert.assertEquals(7, actualResult, DELTA);
+    }
+
+    @Test(expected = BadOperandException.class)
+    public void getResultWithBadOperand() throws BadOperandException {
+        calculator.calculate(5, 3, '9');
+    }
+
+    @Test
+    public void getResultWithMultiplicationWithNull() throws BadOperandException {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/HelloWorldTest.java:83]: add more tests for multiplication

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 14 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java]: done

File: src/test/java/core/basesyntax/HelloWorldTest.java (hunk 15 above)
Comment (voloshynamaryna11) [reply src/test/java/core/basesyntax/HelloWorldTest.java]: done

**Review by KseniiaMakarova — APPROVED** [review]
